File: lintcode/120.word-ladder.py
class Solution:
    """
    @param: start: a string
    @param: end: a string
    @param: dict: a set of string
    @return: An integer
    """

    def ladderLength(self, start, end, dict):
        dict.add(end)
        import queue
        q = queue.Queue()

        q.put((start, 1))
        explored = set()

        while not q.empty():
            curr_str, level = q.get()
            if curr_str == end:
                return level

            import string
            for i in range(len(curr_str)):
                for c in list(string.ascii_lowercase):
                    new_str = curr_str[0:i] + c + curr_str[i + 1:]
                    if new_str in dict and new_str not in explored:
                        q.put((new_str, level + 1))
                        explored.add(new_str)


if __name__ == '__main__':
    res = Solution().ladderLength('hit', 'cog',
                                  set(["hot", "dot", "dog", "lot", "log"]))
    print(res)

File: lintcode/s/120.word-ladder.cpp
#include <queue>
using namespace std;

class Solution {
public:
  /*
   * @param start: a string
   * @param end: a string
   * @param dict: a set of string
   * @return: An integer
   */
  int ladderLength(string &start, string &end, unordered_set<string> &dict) {
    dict.insert(end);

    queue<pair<string, int>> q;
    q.push({start, 1});

    while (!q.empty()) {
      auto p = q.front();
      q.pop();

      string str = p.first;
      cout << str << endl;
      int level = p.second;

      if (str == end) {
        return level;
      }

      string newStr{str};
      for (int i = 0; i < newStr.size(); ++i) {
        for (char ch = 'a'; ch <= 'z'; ++ch) {
          if (str[i] != ch) {
            newStr[i] = ch;
            if (dict.count(newStr)) {
              q.push({newStr, level + 1});
              dict.erase(newStr);
            }
          }
        }
        newStr[i] = str[i];
      }
    }

    return 0;
  }
};
